crates/ruff_linter/src/rules/flake8_type_checking/rules/runtime_string_union.rs

```diff
@@ -57,39 +85,204 @@ pub(crate) fn runtime_string_union(checker: &mut Checker, expr: &Expr) {
+                true
+            } else {
+                strings.push(StringResult {
+                    string: expr,
+                    quotes_are_removable: false,
+                });
+                false
+            }
         }
         Expr::BytesLiteral(_) => {
-            strings.push(expr);
+            strings.push(StringResult {
+                string: expr,
+                quotes_are_removable: false,
+            });
+            false
         }
         Expr::BinOp(ast::ExprBinOp {
             left,
             right,
             op: Operator::BitOr,
             ..
         }) => {
-            traverse_op(left, strings);
-            traverse_op(right, strings);
+            // we don't want short-circuiting here, since we need to collect
+            // string results from both branches
+            traverse_op(semantic, locator, left, strings, settings)
+                & traverse_op(semantic, locator, right, strings, settings)
         }
-        _ => {}
+        _ => true,
     }
 }
+
+/// Traverses the type expression and checks if the expression can safely
+/// be unquoted
+fn quotes_are_removable(semantic: &SemanticModel, expr: &Expr, settings: &LinterSettings) -> bool {
+    match expr {
+        Expr::BinOp(ast::ExprBinOp {
+            left, right, op, ..
+        }) => {
+            match op {
+                Operator::BitOr => {
+                    if settings.target_version < PythonVersion::Py310 {
+                        return false;
+                    }
+                    quotes_are_removable(semantic, left, settings)
+                        && quotes_are_removable(semantic, right, settings)
+                }
+                // for now we'll treat uses of other operators as unremovable quotes
+                // since that would make it an invalid type expression anyways. We skip
+                // walking subscript
+                _ => false,
+            }
+        }
+        Expr::Starred(ast::ExprStarred {
+            value,
+            ctx: ExprContext::Load,
+            ..
+        }) => quotes_are_removable(semantic, value, settings),
+        // Subscript or attribute accesses that are valid type expressions may fail
+        // at runtime, so we have to assume that they do, to keep code working.
+        Expr::Subscript(_) | Expr::Attribute(_) => false,
+        Expr::List(ast::ExprList { elts, .. }) | Expr::Tuple(ast::ExprTuple { elts, .. }) => {
+            for elt in elts {
+                if !quotes_are_removable(semantic, elt, settings) {
+                    return false;
+                }
+            }
+            true
+        }
+        Expr::Name(name) => {
+            semantic.lookup_symbol(name.id.as_str()).is_none()
+                || semantic
+                    .simulate_runtime_load(name, TypingOnlyBindingsStatus::Disallowed)
+                    .is_some()
+        }
+        _ => true,
+    }
+}
```

Comment:
While there is some code duplication between this and `quotes_are_unremovable`, this version is simplified for lookups from a runtime context and needs to use `lookup_symbol` instead of `resolve_name` since the forward reference has not been traversed yet by the checker.

So I'm torn about whether to merge this with `quotes_are_unremovable` into a common function in `helpers.rs` and leveraging an enum with two variants to determine whether we can use `resolve_name` or need to use `lookup_symbol` instead. Alternatively we could try passing in a closure. But both approaches seem kind of messy to me.

We could also say we're fine with the additional overhead of `lookup_symbol` over `resolve_name` and just always use that.